File: day_05/solution.py
import time
start_time = time.time()
def create_stacks(txt):
    numStacks = int(txt[-1].replace(' ', '')[-1])
    stacks = [[] for i in range(numStacks)]

    for i in range(len(txt)-1):
        rIdx = numStacks-2-i
        row = txt[rIdx]
        row = [row[c:c+4][1] for c in range(0, len(row), 4)]
        for sIdx, val in enumerate(row):
            if val != ' ':
                stacks[sIdx].append(val)
    return stacks


def get_top_rearrange_stacks(stacks: list[list[str]], method: callable):
    for line in procedure:
        _, num, _, from_stack, _, to_stack = line.split()
        method(stacks, int(num), int(from_stack), int(to_stack))
    return ''.join([stack[-1] for stack in stacks])


def move_crates(stacks: list[list[str]], num, from_stack, to_stack):
    for _ in range(num):
        stacks[to_stack-1].append(stacks[from_stack-1].pop())


def move_multiple_crates(stacks: list[list[str]], num, from_stack, to_stack):
    stacks[to_stack-1].extend(stacks[from_stack-1][-num:])
    stacks[from_stack-1] = stacks[from_stack-1][:-num]


with open('input.txt') as f:
    stacks_txt, procedure = [txt.splitlines() for txt in f.read().split('\n\n')]

stacks = create_stacks(stacks_txt)
ans = get_top_rearrange_stacks(stacks, move_crates)
print(f"part 1 answer: {ans}")

stacks = create_stacks(stacks_txt)
ans = get_top_rearrange_stacks(stacks, move_multiple_crates)
print(f"part 2 answer: {ans}")
end_time = time.time()
print(f"Time: {round((end_time-start_time)*1000, 4)}ms")

File: day_05/solution.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
	"unicode"
)

type stack[T any] struct {
    Push func(T)
    Pop func() T
    Length func() int
	Values func() []T
	Last func() T
}

func Stack[T any]() stack[T] {
    slice := make([]T, 0)
    return stack[T]{
        Push: func(i T) {
            slice = append(slice, i)
        },
        Pop: func() T {
            res := slice[len(slice)-1]
            slice = slice[:len(slice)-1]
            return res
        },
        Length: func() int {
            return len(slice)
        },
		Values: func() []T {
			return slice
		},
		Last: func() T {
			return slice[len(slice)-1]
		},
    }
}

func main() {
	start_time := time.Now()
	file, err  := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	data := strings.Split(string(file), "\n\n")
	
	moves := strings.Split(data[1], "\n")
	crane := Stack[rune]()
	
	// Part 1
	stacks := parse_stacks(data[0])
	for _, m := range  moves {
		var num_crates, from, to int
		fmt.Sscanf(m, "move %d from %d to %d", &num_crates, &from, &to)
		for i := 0; i < num_crates; i++ {
			move_crates(stacks, crane, 1, from-1, to-1)
		}
	}
	ans1 := ""
	for _, s := range stacks {
		ans1 += string(s.Last())
	} 
	fmt.Printf("Part 1 answer: %s\n", ans1)

	// Part 2
	stacks = parse_stacks(data[0])
	for _, m := range  moves {
		var num_crates, from, to int
		fmt.Sscanf(m, "move %d from %d to %d", &num_crates, &from, &to)
		move_crates(stacks, crane, num_crates, from-1, to-1)
	}
	ans2 := ""
	for _, s := range stacks {
		ans2 += string(s.Last())
	} 
	fmt.Printf("Part 2 answer: %s\n", ans2)
	elapsed := time.Since(start_time)
	fmt.Printf("Time: %s\n", elapsed)
}


func move_crates(stacks []stack[rune], crane stack[rune], num_crates int, from int, to int) {
	// crane := Stack[rune]()
	for i := 0; i < num_crates; i++ {
		crane.Push(stacks[from].Pop())
	}
	for i := 0; i < num_crates; i++ {
		stacks[to].Push(crane.Pop())
	}
}


func parse_stacks(raw string) []stack[rune] {
	data := strings.Split(raw, "\n")
	raw_stack := data[:len(data)-1]
	last_row := strings.Fields(data[len(data)-1])
	num_stacks, _ := strconv.Atoi(last_row[len(last_row)-1])
	stacks := make([]stack[rune], num_stacks)

	for i := range stacks {
		stacks[i] = Stack[rune]()
	}
	for i := len(raw_stack)-1; i >= 0; i-- {
		row := raw_stack[i]
		for j, r := range row {
			column := j / 4
			if unicode.IsLetter(r) {
				stacks[column].Push(r)
			}
		}
	}
	return stacks
}